### DISCUSSION TITLE
details on virtual memory limitss

### DISCUSSION THREAD

**Discussion by bsutton** [isar/isar#982] [Help]
I'm looking to build a cross platform file system in Dart that in particular works in a browser.

I'm looking to use Isar to  to store the db to get around 'in browser' storage limits.
Implementations for other platforms is more a convivence rather than a primary goal.

The db will contain two object types, file and page.
Each file will have a single 'file' object and multiple page objects.
The page objects will be used to store the content of the file (e.g 16KB of data per page)

The api will allow a user to use the equivalent of dart's RandomAccessFile (RAF) to seek to and read any location in the file.

Under the hood the intent is to have the RAF class 'page' the page objects into memory as required. 
A few pages would be retained in memory (either side of the current seek position) and then deleted from memory once the seek
point moved on.
This method should allow a user to read/write multi gigabyte files in a browser without running the browser out of memory and not require the code to request a user's permission to store persistent data.

The Isar documentation states:

> The maximum size of the database file in MiB. Isar uses virtual memory which is not an endless resource so be mindful with the value here. If you open multiple instances they share the available virtual memory so each instance should have a smaller maxSizeMib . 

I'm looking for some clarity on how Isar uses memory.

I assume that the web implementation is different from other platforms.

For non-web systems:
Are the objects mapped via a memory mapped file?
Hence does Isar simply rely on the OS to manage the paging and as such does no explicit object management?
If it does object management can you provide some details on the implementation?

For the web platform.
Can you confirm that each Isar object is mapped to a  indexed db object?
How is memory managed when larges numbers of objects are read?

My memory of Hive is that when you opened a Box it loaded every object into memory. I'm hoping that this is not the case for Isar.

Are there specific problems I should be aware of given the scheme I'm looking to employ?

**Comment by zam101** [comment]
i've built a windows app to manage clients and services (appointments, visa,tickets,hotel) for travel agencies and now when i am about to sell it i discovered that the default max size of db is 2 gb , why cant we just use the max storage available on the machine ? and if i set the max size to very big threshold say 20gb would that be a good solution to avoid app crash?

**Reply by simc** [reply]
@zam101 on desktop machines you can use a very high value without any cost. On phones 2GB is a good default limit

**Reply by zam101** [reply]
ok then i will set it to 50gb and maybe check for the size of  db file at app start-up and ask the user to delete old data if file size is close to that value .
Thank you so much

**Reply by huntrist** [reply]
What would happen, if I set a high number on phone? Would the app crash? What is the cost?
How do I get the max possible number I can set for a given device?
2GB isn't really enough for my use case.

**Reply by annymosse** [reply]
You can partition : 
1- Vertically : by splitting your data into multiple database's files 4 GB max for each Collection | table, and link them by foreign keys (without Isar links), this will reduce your Isar performance little bit while keep your data able to scale much better. (look at the downside NOTE)
2- Horizontally : by making a backup for your data which is older than your user need into chunks by your own custom mechanism, then load it whenever your user query by date outside your current live data date's range (ex: each database's file contains data created before 4 years) and load only that specific database's backup file.
3- Both <Vertically and horizontally>: by mixing both techniques (which much more harder for maintenance and DX developer experience) .

 **NOTE:** currently as today test I found that Isra generator will generate conflicts by providing Collections not belonging to separated database's files example:

```dart
/// models/user.dart
import 'package:isar/isar.dart';

import 'cours.dart';

part 'user.g.dart';

@collection
class User {
  User({
    required this.username,
    required this.firstName,
    required this.lastName,
  });

  Id id = Isar.autoIncrement;

  @Index()
  late String username;
  @Index()
  late String firstName;
  @Index()
  late String lastName;

  late List<int> courses = [];
}
// ---------------------------------------- EOF ----------------------------------------

// models/cours.dart
import 'package:isar/isar.dart';

part 'cours.g.dart';

@collection
class Cours {
  Cours({
    required this.name,
  });

  Id id = Isar.autoIncrement;

  @Index()
  late String name;
  DateTime? createdAt;
}
// ---------------------------------------- EOF ----------------------------------------

/// main.dart
import 'package:isar/isar.dart';

import 'models/cours.dart';
import 'models/user.dart';

void main(List<String> args) async {
  await Isar.initializeIsarCore(download: true);

  final userDb = await Isar.open(
    inspector: false,
    name: "UserDB",
    [
      UserSchema,
    ],
    directory: "/tmp/",
  );
  final coursDb = await Isar.open(
    inspector: false,
    name: "CoursDB",
    [
      CoursSchema,
    ],
    directory: "/tmp/",
  );

  final courses = [
     Cours(name: "cours 1"),
     Cours(name: "cours 2"),
     Cours(name: "cours 3"),
     Cours(name: "cours 4"),
  ];
  final List<int> coursesIds = await coursDb.writeTxn(() async {
     return await coursDb.cours.putAll(courses);
  });
  print("coursDb.cours.count: ${await coursDb.cours.count()}");

  await userDb.writeTxn(() async {
     await userDb.users.put(
       User(
         username: "userName",
         firstName: "firstName",
         lastName: "lastName",
       )..courses.add(coursesIds[1]),
     );
  });

  print("userDb.users.count: ${await userDb.users.count()}");
  final usr = (await userDb.users.get(1))!;
  print("userDb.users[0].id: ${usr.id}");
  print("userDb.users[0].firstName: ${usr.firstName}");
  print("userDb.users[0].lastName: ${usr.lastName}");
  print("userDb.users[0].username: ${usr.username}");
  print("userDb.users[0].courses: ${usr.courses}"); // <== The stored foreign keys 

   // Try to uncomment the following lines then use the autocomplete
   // then you'll find users collection under coursDb !!
  // print("🤔 hmmm ${await coursDb.u} ");
  // print("🤔 hmmm ${await userDb.c} "); // <== same here you'll find courses collection under userDb !!
}
// ---------------------------------------- EOF ----------------------------------------
```

**Comment by gintominto5329** [comment]
Please have a look at [this issue at dart-lang/sdk](https://github.com/dart-lang/sdk/issues/50849), which states that 

> storing just a single `ByteBuffer` of length 256, takes 637 bytes, in IndexedDB

And that is for `ByteBuffer`(the `ArrayBuffer`), which is the most optimized and efficient type, on the web, paging would be a disaster, and 2G limit, would take us to around 700M, if paging is used,

Nothing concrete, but feasibility analysis should help,

Thanks

**Reply by bsutton** [reply]
So I'm speculating that the over head if fixed which means that if we increase the page size we reduce over head as a percentage.

On a 256 byte buffer that expands to 637 bytes the over head is 60%.
If we use a page size of 16K then the over head is 2%.
So on a 2GB allocation we lose 40MB yielding 1960MB usable space.

**Reply by simc** [reply]
None of this is relevant for the native Dart/Rust since IndexedDB is only used in the browser. And there is no memory limit for IndexedDB This discussion is not going in the right direction

**Reply by bsutton** [reply]
These are the question that I need in order to understand how isar works.


I'm looking for some clarity on how Isar uses memory.

I assume that the web implementation is different from other platforms.

For non-web systems:
Are the objects mapped via a memory mapped file?
Hence does Isar simply rely on the OS to manage the paging and as such does no explicit object management?
If it does object management can you provide some details on the implementation?

For the web platform.
Can you confirm that each Isar object is mapped to a indexed db object?
How is memory managed when larges numbers of objects are read?

My memory of Hive is that when you opened a Box it loaded every object into memory. I'm hoping that this is not the case for Isar.

**Reply by simc** [reply]
Isar Native uses a memory mapped file and mostly relies on OS paging. Since iOS specifically limits the available virtual memory the max database file is limited by that. Same as Realm and ObjectBox.

The web implementation in v2.5.0 does whatever IndexedDB does.

The new web implementation that will be available in Isar 4 will use a different mechanism. Details  tbd.
